contracts/v2/accounts/default/implementation_default.sol

```diff
@@ -119,6 +171,61 @@ contract Record is Constants {
     ) external returns (bytes4) {
         return 0xbc197c81; // bytes4(keccak256("onERC1155BatchReceived(address,address,uint256[],uint256[],bytes)"))
     }
+
+    /**
+     * @dev Marks a bytes32 `message` (signature digest) as signed, making it verifiable by EIP-1271 `isValidSignature()`.
+     * @param message data hash to be allow-listed as signed. Input `message` is hashed with `domainSeparatorV4()` according to EIP712 typed data (`EIP1271_TYPE_HASH`)
+    */
+    function signMessage(bytes32 message) external {
+        require(msg.sender == address(this), "not-self");
+
+        // hashing with domain separator mitigates any potential replaying on other networks or other Avocados of the same owner
+        message = ECDSA.toTypedDataHash(
+            domainSeparatorV4(),
+            keccak256(abi.encode(EIP1271_TYPE_HASH, message))
+        );
+
+        _signedMessages[message] = true;
+
+        emit LogSignedMessage(message);
+    }
+
+    /**
+     * @dev Removes a previously `signMessage()` signed bytes32 `message_` (signature digest).
+     * @param message data hash to be removed from allow-listed signatures
+    */
+    function removeSignedMessage(bytes32 message) external {
+        require(msg.sender == address(this), "not-self");
+
+        delete _signedMessages[message];
+
+        emit LogRemoveSignedMessage(message);
+    }
+
+    /**
+     * @dev Should return whether the signature provided is valid for the provided data
+     * @param hash      Hash of the data to be signed
+     * @param signature Signature byte array associated with data
+    */
+    function isValidSignature(bytes32 hash, bytes memory signature) external view returns (bytes4 magicValue) {
+        // hashing with domain separator mitigates any potential replaying on other networks or other Avocados of the same owner
+        hash = ECDSA.toTypedDataHash(
+            domainSeparatorV4(),
+            keccak256(abi.encode(EIP1271_TYPE_HASH, hash))
+        );
+
+        if (signature.length == 0) {
+            // must be pre-allow-listed via `signMessage()` method
+            require(_signedMessages[hash], "invalid-signed-message");
+        } else {
+            address signer = ECDSA.recover(hash, signature);
+
+            require(_auth[signer], "invalid-EIP-1271-signature");
+        }
+
+        return EIP1271_MAGIC_VALUE;
```

Comment:
Also, need to hash DSA address with hash. Right?

Reply:
https://github.com/Instadapp/dsa-contracts/pull/80/files#r1681020213